### ISSUE TITLE
Ledger: Define authentication flows

### ISSUE THREAD

**Issue by landitus** [blockstack/blockstack-app#2355]
When connecting to dApps, we need to go through the Ledger connection flow. There are technical limitations to the current Authentication flow, so we are proposing a new method for connecting with Ledger (and we plan to deprecate the current Auth code).

In order to build the Ledger authentication flow, we should cover the following:
- [x] Map the Authentication and [Request credentials](https://github.com/hirosystems/stacks-wallet-web/issues/2371#issuecomment-1102535366) flows with Ledger 
- [x] Prototype in Figma
- [ ] Define tracking events

**Comment by landitus** [comment]
Here's a draft of the User flows for the **Connect wallet** feature:
- It includes the soon to be deprecated (and current) authentication flow
- Adds a new flow for soon to be added **Request Accounts** method

[Figma](https://www.figma.com/file/o0VSeFzhYI2GgnDcX0NPmU/User-lifecycle-flow?node-id=871%3A983)

<img width="4512" alt="User lifecycle flow - Connect Wallet" src="https://user-images.githubusercontent.com/239215/164510089-8213d5a0-f760-4f2e-9274-dc719d82b6cf.png">

I will draft a prototype based on this flow so we can review it. cc/ @kyranjamie @markmhx

**Comment by landitus** [comment]
In the case a user tries to connect with Ledger to an application using the Authentication method, which is not supported, should we display an error message?

<img width="872" alt="Screenshot 2022-04-21 at 18 46 30" src="https://user-images.githubusercontent.com/239215/164510503-f8965053-da4b-472e-916b-56f55dea9f86.png">

**Comment by landitus** [comment]
Based on feedback, the Connect wallet flow is now much more simple because there's no need to involve the Ledger during the account selection flow anymore. 

We changed the branching decision between Auth and Request account. Now its based on wether the dApp needs Authentication or just the public keys.

[Figma](https://www.figma.com/file/o0VSeFzhYI2GgnDcX0NPmU/User-lifecycle-flow?node-id=871%3A983)

<img width="4512" alt="User lifecycle flow - Connect Wallet - v2" src="https://user-images.githubusercontent.com/239215/164749366-9e574c71-bccc-4a81-907f-67b931035f2f.png">

**Comment by landitus** [comment]
There's one possible flow we are not covering, that can be triggered when the user is on a dApp and clicks Connect Wallet (and the Hiro wallet has not yet been setup). This seems part edger case, part something totally feasible, right?

- Do we have any telemetry that would allow us to understand how frequent a setup happens in this scenario (for software wallets)?
- For the Auth v1 flow, we should block the flow if the user chooses Ledger (allowing them to go back and choose a software wallet
- For the Request accounts flow, I'm not sure what we should do. The whole Ledger setup happens inside the Connect window (check screenshot). Maybe it's a good thing it all happens without making the user go to a new tab? Still, I don't know what would happen once the Ledger is actually connected. 
- We could do nothing and add a tracking event and create report to see how many setup Ledger from the Connect window.

Wdyt @markmhx @kyranjamie ?

Here's a video on how the flow looks with the current build.

https://user-images.githubusercontent.com/239215/164749992-0f24e687-eb2f-4baf-a0ae-944ca05fe2bd.mp4

A simulation of what the user would see when Connecting though an app a brand new Wallet
<img width="1440" alt="simulation" src="https://user-images.githubusercontent.com/239215/164751949-e9b21bff-2015-4c2a-a010-c5850f9c48fd.png">

**Comment by kyranjamie** [comment]
> Do we have any telemetry that would allow us to understand how frequent a setup happens in this scenario?

I don't think so

> For the Auth v1 flow, we should block the flow if the user chooses Ledger (allowing them to go back and choose a software wallet

A warning/tooltip, at least, would be nice

> For the Request accounts flow, I'm not sure what we should do. I think the Ledger connect flow is not great due to space constrains. Still, I don't know what would happen once the Ledger is actually connected. Maybe it's a good thing it all happens without making the user go to a new tab?

Umm 🤨 . It think it might be possible to go off to full screen mode and return to the popup following onboarding. This wouldn't be super trivial, though.

**Comment by landitus** [comment]
Thanks @kyranjamie. Will try these solutions first and see how they look

**Comment by landitus** [comment]
There's a more in depth solution for connecting an dapp to a brand new Wallet (without account created) [here](https://github.com/hirosystems/stacks-wallet-web/pull/2242#issuecomment-1109785561)

**Comment by landitus** [comment]
I've updated the prototype based on [this proposal](https://github.com/hirosystems/stacks-wallet-web/pull/2242#issuecomment-1109785561) and with help of @kyranjamie. This prototype shows authenticated and non-authenticated scenarios with both **Auth** and **Request accounts** methods, from the point of view of a user interacting with a dApp.

**Here are the new updates:**
- We've decided to force the onboarding (or resuming the onboarding if abandoned) always in a new tab. This way, Ledger can be set up without issues
- We are now accounting for resuming the onboarding, if the user installed the Wallet but didn't create the first account. We are proposing a new screen to be displayed in the Wallet pop-up on top of the app: "You still need to setup  your Hiro Wallet"
- As an enhancement, once the Wallet setup is complete, we are proposing a callback to return to the app. This will only be included if the Wallet setup originated from an dAapp.

**Question**
My only concern is [this scenario](https://www.figma.com/proto/VQPOPkkrfVcf0BEd9kz3HZ/%232275---Ledger%3A-Onboarding-%2B-Transaction-signing?page-id=1341%3A39025&node-id=1351%3A36334&viewport=241%2C48%2C0.25&scaling=scale-down&starting-point-node-id=1351%3A36334&show-proto-sidebar=1) where the user could technically go into STXNFT (or any site with the Auth method) to buy an NFT, setup a brand new Ledger wallet and realize only at the end of the whole process, that Ledger is not supported because of the old Auth method. I know we are waiting on fix on this, so mind this is reflecting the current Auth method situation cc/ @markmhx 

[Check the Figma prototype](https://www.figma.com/proto/VQPOPkkrfVcf0BEd9kz3HZ/%232275---Ledger%3A-Onboarding-%2B-Transaction-signing?page-id=1341%3A39025&node-id=1442%3A40518&viewport=241%2C48%2C0.25&scaling=scale-down&starting-point-node-id=1442%3A40518&show-proto-sidebar=1), where each use-case/flow has a flowchart as well.

<img width="1440" alt="Choose an account" src="https://user-images.githubusercontent.com/239215/166733638-efc8283c-c64b-4afe-99f7-5f5ddc39d11a.png">
<img width="1440" alt="Wallet home" src="https://user-images.githubusercontent.com/239215/166734284-e3ece57b-a8a3-41a8-ab62-751611619af9.png">

**Comment by markmhendrickson** [comment]
> We are now accounting for resuming the onboarding, if the user installed the Wallet but didn't create the first account. We are proposing a new screen to be displayed in the Wallet pop-up on top of the app: "You still need to setup  your Hiro Wallet"

Is this screen to be shown specifically when the user tries to sign a transaction having not yet set up their wallet yet? Or are there other cases in which this is shown?

> As an enhancement, once the Wallet setup is complete, we are proposing a callback to return to the app. This will only be included if the Wallet setup originated from an dAapp.

We last looked into this a couple of months ago upon which @fbwoolf [flagged some difficulties](https://github.com/hirosystems/stacks-wallet-web/issues/1808#issuecomment-1068125021) and @landitus [was going to investigate Metamask's approach further](https://github.com/hirosystems/stacks-wallet-web/issues/1808#issuecomment-1068151113), if I'm not mistaken.

Do we have a better sense these days of how we'd go about this?

> My only concern is [this scenario](https://www.figma.com/proto/VQPOPkkrfVcf0BEd9kz3HZ/%232275---Ledger%3A-Onboarding-%2B-Transaction-signing?page-id=1341%3A39025&node-id=1351%3A36334&viewport=241%2C48%2C0.25&scaling=scale-down&starting-point-node-id=1351%3A36334&show-proto-sidebar=1) where the user could technically go into STXNFT (or any site with the Auth method) to buy an NFT, setup a brand new Ledger wallet and realize only at the end of the whole process, that Ledger is not supported because of the old Auth method. I know we are waiting on fix on this, so mind this is reflecting the current Auth method situation

Our current plan is to release Ledger support _with_ app authentication support, leveraging Zondax's upgrade to the Stacks app that enables this. So, Ledger should be supported fine in this STXNFT case?

**Comment by landitus** [comment]
> Is this screen to be shown specifically when the user tries to sign a transaction having not yet set up their wallet yet? Or are there other cases in which this is shown?

Yes, and also if they try to connect and authenticaticate with a wallet that was not set up.

> Do we have a better sense these days of how we'd go about this?

I forgot about looking into MetaMask. But I'll take a look!

> Our current plan is to release Ledger support with app authentication support, leveraging Zondax's upgrade to the Stacks app that enables this. So, Ledger should be supported fine in this STXNFT case?

If it's supported fine, there would be no issues. Would you like me to update the mocks assuming we get Ledger supported?

**Comment by markmhendrickson** [comment]
> Yes, and also if they try to connect and authenticaticate with a wallet that was not set up.

For both transaction signing and authenticate, clicking "Setup Hiro Wallet" goes to the landing page ("Explore world of Stacks") with options to generate key or sign in?

If so, perhaps it's best to send the user directly there and simply show a message on the top saying "Please set up your wallet by choosing an option below before proceeding to [sign into Testing App|sign a transaction with Testing App]". That way we save the user a step and save ourselves the need for an entirely new screen.

wdyt?

> If it's supported fine, there would be no issues. Would you like me to update the mocks assuming we get Ledger supported? 

Yep I believe so. @kyranjamie can you confirm?

> I forgot about looking into MetaMask. But I'll take a look!

🙏

**Comment by kyranjamie** [comment]
> We last looked into this a couple of months ago upon which @fbwoolf https://github.com/hirosystems/stacks-wallet-web/issues/1808#issuecomment-1068125021

This referred to handling the _returning to app scenario_ when user initially _installs_ the wallet. That is not presented here. This flow describes wallet installed but _not_ set up yet scenario.

> Our current plan is to release Ledger support with app authentication support, leveraging Zondax's upgrade to the Stacks app that enables this. So, Ledger should be supported fine in this STXNFT case?

This change is not necessarily a golden ticket to full hardware wallet support. I'm not sure yet if auth alone without an `appPrivateKey` is that useful for many dApps. This is my next priority to figure out, in testing the new Ledger app version.

**Comment by markmhendrickson** [comment]
> This referred to handling the returning to app scenario when user initially installs the wallet. That is not presented here. This flow describes wallet installed but not set up yet scenario.

Ah I see, got it. Should we break this out into its own issue, separate from Ledger since it's not strictly needed for it?

Same question for the new setup screen or message upon auth or signing for users who haven't completed setup. It seems we could tackle these as a non-Ledger-dependent enhancement.

> This is my next priority to figure out, in testing the new Ledger app version.

Let's test ASAP so we can get clarity on this.

**Comment by landitus** [comment]
> Ah I see, got it. Should we break this out into its own issue, separate from Ledger since it's not strictly needed for it?

Sounds good to me

> Same question for the new setup screen or message upon auth or signing for users who haven't completed setup. It seems we could tackle these as a non-Ledger-dependent enhancement.

This one is intended for the Ledger setup UX. [As we've identified here](https://github.com/hirosystems/stacks-wallet-web/pull/2242#issuecomment-1109785561), we'd like to route all onboarding flows to a new tab, especially as Ledger is better set up that way. We could do a quick fix for this (and delay building this new screen) if we set the "I have a Ledger" link to open in a new tab. This way, at least, Ledger is not set up in the pop-up window, but in a full tab.

<img width="523" alt="165317818-5016c914-cb40-45a6-9b06-d396773e8120" src="https://user-images.githubusercontent.com/239215/167183214-a0063461-cff7-4359-a3b5-d930293c30c9.png">

**Comment by markmhendrickson** [comment]
> We could do a quick fix for this (and delay building this new screen) if we set the "I have a Ledger" link to open in a new tab. This way, at least, Ledger is not set up in the pop-up window, but in a full tab.

Yep, this is what I'd suggest here in support of the initial release of Ledger. 

I think we also discussed converting the Ledger connection overlay below into its own page view so we can avoid loading the "Explore the world of Stacks" landing page first, only to display the overlay immediately on top of it in this scenario. That sounds like a good enhancement to me as well for this initial release, assuming it doesn't require a lot of code change.

<img width="391" alt="image" src="https://user-images.githubusercontent.com/28991/167351675-67f25d00-2f7e-4d76-b839-f1a47f02b4f4.png">
